ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/NoUnusedImportsRule.kt

```diff
@@ -44,18 +45,39 @@ class NoUnusedImportsRule : Rule("no-unused-imports") {
         // by (https://github.com/shyiko/ktlint/issues/54)
         "getValue", "setValue"
     )
+
+    private val conditionalWhitelist = mapOf<String, (node: ASTNode) -> Boolean>(
+        Pair(
+            // Ignore provideDelegate if there is a `by` anywhere in the file
+            "org.gradle.kotlin.dsl.provideDelegate",
+            { rootNode ->
+                var hasByKeyword = false
+                rootNode.visit { child ->
+                    if (child.elementType == BY_KEYWORD) {
+                        hasByKeyword = true
+                        return@visit
+                    }
+                }
+                hasByKeyword
+            }
+        )
+    )
+
     private val ref = mutableSetOf<String>()
     private val imports = mutableSetOf<String>()
     private var packageName = ""
+    private var rootNode: ASTNode? = null
 
     override fun visit(
         node: ASTNode,
         autoCorrect: Boolean,
         emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit
     ) {
         if (node.isRoot()) {
+            rootNode = node
             ref.clear() // rule can potentially be executed more than once (when formatting)
             ref.add("*")
+            imports.clear()
```

Comment:
This is an unrelated change, but based on the comment on Line 78, it seems necessary.